spec/eofv0_verkle.md

```diff
@@ -105,6 +105,158 @@ The same as above except encode the values as 6-bit numbers
+
+This is beneficial if our assumption is correct that most contracts only contain a limited number
+of offending cases. Our initial analysis suggests this is the case, e.g. Uniswap router has 9 cases,
+one of the Arbitrum validator contracts has 6 cases.
+
+Since Solidity contracts have a trailing metadata, which contains a Keccak-256 (32-byte) hash of the
+source, there is a 12% probability ($1 - (255/256)^{32}$) that at least one of the bytes of the hash
+will contain the `0x5b` value, which gives our minimum probability of having at least one invalid
+`JUMPDEST` in the contract.
+
+Let's create a map of `invalid_jumpdests[chunk_no] = first_instruction_offset`. We can densely encode this
+map using techniques similar to *run-length encoding* to skip distances and delta-encode offsets.
+This map is always fully loaded prior to execution, and so it is important to ensure the encoded
+version is as dense as possible (without sacrificing on complexity).
+
+In *scheme 1*, for each entry in `invalid_jumpdests`:
+- 1-bit mode (`skip`, `value`)
+- For skip-mode:
+  - 10-bit number of chunks to skip
+- For value-mode:
+  - 6-bit `first_instruction_offset`
+
+Worst case encoding where each chunk contains an invalid `JUMPDEST`:
+```
+total_chunk_count = 24576 / 32 = 768
+total_chunk_count * (1 + 6) / 8 = 672 # bytes for the header, i.e. 2.7% overhead
+number_of_verkle_leafs = total_chunk_count / 32 = 21
+```
+
+*Scheme 2* differs slightly:
+- 1-bit mode (`skip`, `value`)
+- For skip-mode:
+  - 10-bit number of chunks to skip
+- For value-mode:
+  - 4-bit number of chunks to skip
+  - 6-bit `first_instruction_offset`
+
+Worst case encoding:
+```
+total_chunk_count = 24576 / 32 = 768
+total_chunk_count * (1 + 4 + 6) / 8 = 1056 # bytes for the header, i.e. 4.1% overhead
+number_of_verkle_leafs = total_chunk_count / 32 = 33
+```
+
+The decision between *scheme 1* and *scheme 2*, as well as the best encoding sizes, can be determined
+through analysing existing code.
+
+#### Header location
+
+It is possible to place above as part of the "EOFv0" header, but given the upper bound of number of chunks occupied is low (33 vs 21),
+it is also possible to make this part of the Verkle account header.
+
+This second option allows for the simplification of the `code_size` value, as it does not need to change.
+
+#### Runtime after Verkle
+
+During runtime execution two checks must be done in this order:
+1) Check if the destination is on the invalid list, and abort if so.
+2) Check if the value in the chunk is an actual `JUMPDEST`, and abort if not.
+
+It is possible to reconstruct sparse account code prior to execution with all the submitted chunks of the transaction
+and perform `JUMPDEST`-validation to build up a relevant *valid `JUMPDEST` locations* map instead.
+
+#### Analysis
+
+We have analyzed two contracts, Arbitrum validator and Uniswap router.
+
+Arbitrum (2147-bytes long):
+```
+(chunk offset, chunk number, pushdata offset)
+malicious push byte: 85 2 21
```

Comment:
This analysis is wrong because we have to encode *first instruction offset* instead of *first invalid jumpdest offset*. I think we should remove this section or at least mark is as incorrect until I'll come with proper analysis.